File: javadarts/src/test/java/com/sdoward/barstewarddarts/android/java/GameTest.java
package com.sdoward.barstewarddarts.android.java;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

public class GameTest {

    @Mock
    private Game.GameInstructions instructions;
    private Game game;
    private List<Player> players = new ArrayList<>(2);

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        players.add(new Player("name1", 1));
        players.add(new Player("name2", 2));

        game = new Game(players, instructions);
    }

    @Test
    public void shouldDisplayPlayers() {
        game.start();
        verify(instructions).displayCurrentPlayer("name1");
        game.acceptThrow(Throw.innerBull());
        game.acceptThrow(Throw.innerBull());
        game.acceptThrow(Throw.innerBull());
        verify(instructions).displayCurrentPlayer("name2");
        game.acceptThrow(Throw.innerBull());
        game.acceptThrow(Throw.innerBull());
        game.acceptThrow(Throw.innerBull());
        verify(instructions, times(2)).displayCurrentPlayer("name1");
    }

    @Test
    public void shouldDecrementLifeWhenOwnNumberHit() {
        game.acceptThrow(Throw.single(1));
        game.acceptThrow(Throw.single(1));
        game.acceptThrow(Throw.miss());
        assertThat(game.getPlayers().get(0).getLives()).isEqualTo(1);
    }

    @Test
    public void shouldRemove2LivesWhenOwnNumberHit() {
        game.acceptThrow(Throw.duble(1));
        game.acceptThrow(Throw.miss());
        game.acceptThrow(Throw.miss());
        assertThat(game.getPlayers().get(0).getLives()).isEqualTo(1);
    }

    @Test
    public void shouldRemove3LivesWhenOwnNumberHit() {
        game.acceptThrow(Throw.triple(1));
        game.acceptThrow(Throw.miss());
        game.acceptThrow(Throw.miss());
        assertThat(game.getPlayers().get(0).getLives()).isEqualTo(0);
    }

    @Test
    public void shouldIncrementLifeWhenOtherPlayersNumberHit() {
        game.acceptThrow(Throw.single(2));
        game.acceptThrow(Throw.single(2));
        game.acceptThrow(Throw.miss());
        assertThat(game.getPlayers().get(1).getLives()).isEqualTo(5);
    }

    @Test
    public void shouldAdd2LivesWhenDoubleOtherOtherPlayer() {
        game.acceptThrow(Throw.duble(2));
        game.acceptThrow(Throw.miss());
        game.acceptThrow(Throw.miss());
        assertThat(game.getPlayers().get(1).getLives()).isEqualTo(5);
    }

    @Test
    public void shouldAdd3LivesWhenDoubleOtherOtherPlayer() {
        game.acceptThrow(Throw.triple(2));
        game.acceptThrow(Throw.miss());
        game.acceptThrow(Throw.miss());
        assertThat(game.getPlayers().get(1).getLives()).isEqualTo(6);
    }

    @Test
    public void shouldEndGameWhenAnyLifeIsBelow1() {
        game.acceptThrow(Throw.miss());
        game.acceptThrow(Throw.miss());
        game.acceptThrow(Throw.triple(1));
        verify(instructions).displayWinner("name1");
    }

    @Test
    public void shouldNotEndGameWhenNoLivesAreBelow1() {
        game.acceptThrow(Throw.miss());
        game.acceptThrow(Throw.miss());
        game.acceptThrow(Throw.miss());
        verify(instructions, never()).displayWinner(anyString());
    }

    @Test
    public void shouldDisplayListOfDrinkingInstructions() {
        game.acceptThrow(Throw.triple(1));
        game.acceptThrow(Throw.duble(2));
        game.acceptThrow(Throw.miss());

        List<String> expectingDrinkingInstructions = new ArrayList(2);
        expectingDrinkingInstructions.add("name1 drink 3");
        expectingDrinkingInstructions.add("name2 drink 2");

        verify(instructions).displayDrinkingInstructions(expectingDrinkingInstructions);
    }

}

File: kotlindarts/src/test/kotlin/com/sdoward/barstewarddarts/android/kotlin/GameTest.kt
package com.sdoward.barstewarddarts.android.kotlin

import org.assertj.core.api.Assertions
import org.junit.Test
import org.mockito.Mockito
import org.mockito.Mockito.*

class GameTest {

    private val instructions: GameInstructions = Mockito.mock(GameInstructions::class.java)
    private val players = listOf(Player("name1", 1), Player("name2", 2))
    private var game = Game(players, instructions)

    @Test
    fun shouldDisplayPlayers() {
        game.start()
        verify(instructions).displayCurrentPlayer("name1")
        game.apply {
            acceptThrow(Throw.innerBull())
            acceptThrow(Throw.innerBull())
            acceptThrow(Throw.innerBull())
        }
        verify(instructions).displayCurrentPlayer("name2")
        game.apply {
            acceptThrow(Throw.innerBull())
            acceptThrow(Throw.innerBull())
            acceptThrow(Throw.innerBull())
        }
        verify(instructions, times(2)).displayCurrentPlayer("name1")
    }

    @Test
    fun shouldDecrementLifeWhenOwnNumberHit() {
        game.apply {
            acceptThrow(Throw.single(1))
            acceptThrow(Throw.single(1))
            acceptThrow(Throw.miss())
        }
        Assertions.assertThat(game.players[0].lives).isEqualTo(1)
    }

    @Test
    fun shouldRemove2LivesWhenOwnNumberHit() {
        game.apply {
            acceptThrow(Throw.duble(1))
            acceptThrow(Throw.miss())
            acceptThrow(Throw.miss())
        }
        Assertions.assertThat(game.players[0].lives).isEqualTo(1)
    }

    @Test
    fun shouldRemove3LivesWhenOwnNumberHit() {
        game.apply {
            acceptThrow(Throw.triple(1))
            acceptThrow(Throw.miss())
            acceptThrow(Throw.miss())
        }
        Assertions.assertThat(game.players[0].lives).isEqualTo(0)
    }

    @Test
    fun shouldIncrementLifeWhenOtherPlayersNumberHit() {
        game.apply {
            acceptThrow(Throw.single(2))
            acceptThrow(Throw.single(2))
            acceptThrow(Throw.miss())
        }
        Assertions.assertThat(game.players[1].lives).isEqualTo(5)
    }

    @Test
    fun shouldAdd2LivesWhenDoubleOtherOtherPlayer() {
        game.apply {
            acceptThrow(Throw.duble(2))
            acceptThrow(Throw.miss())
            acceptThrow(Throw.miss())
        }
        Assertions.assertThat(game.players[1].lives).isEqualTo(5)
    }

    @Test
    fun shouldAdd3LivesWhenDoubleOtherOtherPlayer() {
        game.apply {
            acceptThrow(Throw.triple(2))
            acceptThrow(Throw.miss())
            acceptThrow(Throw.miss())
        }
        Assertions.assertThat(game.players[1].lives).isEqualTo(6)
    }

    @Test
    fun shouldEndGameWhenAnyLifeIsBelow1() {
        game.apply {
            acceptThrow(Throw.miss())
            acceptThrow(Throw.miss())
            acceptThrow(Throw.triple(1))
        }
        verify(instructions).displayWinner("name1")
    }

    @Test
    fun shouldNotEndGameWhenNoLivesAreBelow1() {
        game.apply {
            acceptThrow(Throw.miss())
            acceptThrow(Throw.miss())
            acceptThrow(Throw.miss())
        }
        verify(instructions, never()).displayWinner(anyString())
    }

    @Test
    fun shouldDisplayListOfDrinkingInstructions() {
        game.apply {
            acceptThrow(Throw.triple(1))
            acceptThrow(Throw.duble(2))
            acceptThrow(Throw.miss())
        }

        val expectingDrinkingInstructions = mutableListOf<String>()
        expectingDrinkingInstructions.add("name1 drink 3")
        expectingDrinkingInstructions.add("name2 drink 2")

        verify(instructions).displayDrinkingInstructions(expectingDrinkingInstructions)
    }

}